[HUDI-5690] Add simpleBucketPartitioner to support using the simple bucket index under bulkinsert

wuwenchi opened the pull request:
### Change Logs

Add simpleBucketPartitioner to support using the simple bucket index under bulkinsert.

### Impact

none

### Risk level (write none, low medium or high below)

low

### Documentation Update

none

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

wuwenchi commented:
In the code, the part of the abstract class RDDBucketIndexPartitioner comes from RDDConsistentBucketPartitioner. RDDSimpleBucketPartitioner added its own repartitionRecords method.

wuwenchi commented:
@YuweiXiao @leesf @codope can you help review it? Thanks!

danny0405 commented:
@jonvex @lokeshj1703 Do you have interests on reviewing?

jonvex commented:
@danny0405 I did some work with bulkinsertpartitioner so I can take a look

jonvex commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDBucketIndexPartitioner.java:
```diff
@@ -18,15 +18,155 @@
 
 /**
  * Abstract of bucket index bulk_insert partitioner
  * TODO implement partitioner for SIMPLE BUCKET INDEX
```
remove todo line

wuwenchi replied:
Done

jonvex commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDBucketIndexPartitioner.java:
```diff
@@ -18,15 +18,155 @@
+  final List<String> indexKeyFields;
+  final boolean consistentLogicalTimestampEnabled;
+  public final List<Boolean> doAppend = new ArrayList<>();
+  public final List<String> fileIdPfxList = new ArrayList<>();
```
I think these can be private or protected

wuwenchi replied:
Done

YuweiXiao commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDConsistentBucketPartitioner.java:
```diff
@@ -18,82 +18,45 @@
     this(table, Collections.emptyMap(), false);
     ValidationUtils.checkArgument(table.getIndex() instanceof HoodieSparkConsistentBucketIndex,
         "RDDConsistentBucketPartitioner can only be used together with consistent hashing bucket index");
-    ValidationUtils.checkArgument(table.getMetaClient().getTableType().equals(HoodieTableType.MERGE_ON_READ),
```
We could keep this validation, as Consistent Hashing only supports MOR table.

wuwenchi replied:
This validation was moved to RDDBucketIndexPartitioner.

YuweiXiao replied:
Not sure if SIMPLE bucket also has this constraint. 

The reason why we constraint CONSISTENT HASHING to MOR table only, is that some mechanism (i.e., concurrent control) rely on the log feature of MOR table.

YuweiXiao commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bucket/HoodieSimpleBucketIndex.java:
```diff
@@ -44,7 +44,7 @@ public HoodieSimpleBucketIndex(HoodieWriteConfig config) {
   }
 
-  private Map<Integer, HoodieRecordLocation> loadPartitionBucketIdFileIdMapping(
+  public Map<Integer, HoodieRecordLocation> loadPartitionBucketIdFileIdMapping(
```
Maybe keep it private.

wuwenchi replied:
HoodieSimpleBucketIndex and RDDSimpleBucketPartitioner belong to different packages and can only use public.

YuweiXiao replied:
Could it be protected?

YuweiXiao reviewed (approved):
Look good overall!

bvaradar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/storage/HoodieSimpleBucketLayout.java:
```diff
@@ -34,6 +34,7 @@ public class HoodieSimpleBucketLayout extends HoodieStorageLayout {
   public static final Set<WriteOperationType> SUPPORTED_OPERATIONS = CollectionUtils.createImmutableSet(
       WriteOperationType.INSERT,
       WriteOperationType.INSERT_PREPPED,
+      WriteOperationType.BULK_INSERT,
```
@wuwenchi @YuweiXiao : should HoodieBucketIndex.requiresTagging also return True for Bulk_Index ?

wuwenchi replied:
Yes, `BULK_INSERT` here should return true. Fixed.

bvaradar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDBucketIndexPartitioner.java:
```diff
@@ -18,15 +18,155 @@
  */
 public abstract class RDDBucketIndexPartitioner<T>
     implements BulkInsertPartitioner<JavaRDD<HoodieRecord<T>>> {
+
```
HoodieBucketIndex is defined in Engine agnostic way (uses HoodieData and HoodieEngineContext). Can we also define the base partitioner class using these abstractions instead of directly using JavaRDD ?

wuwenchi replied:
This is what was originally written here. I guess it is because flink already has its own bucket implementation, and it involves re-partitioning operations. It is more convenient to use rdd directly.

bvaradar replied:
@wuwenchi : I agree it was written like that before but wanted to take this opportunity to correct it as we are bringing in more engines to this specific use-case. Can you elaborate more why it is convenient to use rdd directly ? One of the main reason why we introduced these abstractions is to have proper reuse of core components across engines while being flexible to extend in engine specific way. Thanks

wuwenchi replied:
It's a good idea, I tried to abstract some methods into hudi-client.

bvaradar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDBucketIndexPartitioner.java:
```diff
@@ -18,15 +18,155 @@
  */
 public abstract class RDDBucketIndexPartitioner<T>
     implements BulkInsertPartitioner<JavaRDD<HoodieRecord<T>>> {
+
```
HoodieBucketIndex is defined in Engine agnostic way (uses HoodieData and HoodieEngineContext). Can we also define the base partitioner class using these abstractions instead of directly using JavaRDD ?

bvaradar reviewed (changes requested):
@wuwenchi : Thanks for the PR. Have few comments.

bvaradar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDSimpleBucketPartitioner.java:
```diff
@@ -0,0 +1,106 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class RDDSimpleBucketPartitioner<T extends HoodieRecordPayload> extends RDDBucketIndexPartitioner<T> {
```
similar class name change to include RDDSimpleBucketBulkInsertPartitioner.

wuwenchi replied:
done

bvaradar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDSimpleBucketPartitioner.java:
```diff
@@ -0,0 +1,106 @@
+        .distinct().collect().stream()
+        .collect(Collectors.toMap(p -> p, p -> {
+          Map<Integer, HoodieRecordLocation> locationMap = index.loadPartitionBucketIdFileIdMapping(table, p);
+          HashMap<Integer, String> fileIdMap = new HashMap<>();
```
rename to bucketIdToFileIdMap

wuwenchi replied:
done

bvaradar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/execution/bulkinsert/TestRDDSimpleBucketPartitioner.java:
```diff
@@ -0,0 +1,115 @@
+    HoodieTestDataGenerator dataGenerator = new HoodieTestDataGenerator();
+    List<HoodieRecord> records = dataGenerator.generateInserts("0", 100);
+    HoodieJavaRDD<HoodieRecord> javaRDD = HoodieJavaRDD.of(records, context, 1);
+    javaRDD.map(HoodieRecord::getPartitionPath).count();
```
this line can be removed.

wuwenchi replied:
done

bvaradar reviewed (changes requested):
Overall looks good. Few comments

bvaradar commented:
@wuwenchi : Can you look at the PR comments and address them when you get a chance.

wuwenchi commented:
> @wuwenchi : Can you look at the PR comments and address them when you get a chance.

@bvaradar  Sorry for the delay... Modified some, but there seems to be a conflict, I will solve it tomorrow

wuwenchi commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bucket/HoodieSimpleBucketIndex.java:
```diff
@@ -44,7 +44,7 @@ public HoodieSimpleBucketIndex(HoodieWriteConfig config) {
   }
 
-  private Map<Integer, HoodieRecordLocation> loadPartitionBucketIdFileIdMapping(
+  public Map<Integer, HoodieRecordLocation> loadPartitionBucketIdFileIdMapping(
```
done

wuwenchi commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/BucketIndexPartitioner.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Abstract of bucket index bulk_insert partitioner
+ */
+public abstract class BucketIndexPartitioner<T> implements BulkInsertPartitioner<T> {
```
done

wuwenchi commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDSimpleBucketPartitioner.java:
```diff
@@ -0,0 +1,106 @@
+  @Override
+  public JavaRDD<HoodieRecord<T>> repartitionRecords(JavaRDD<HoodieRecord<T>> records, int outputPartitions) {
+    HoodieSimpleBucketIndex index = (HoodieSimpleBucketIndex) table.getIndex();
+    HashMap<String, Integer> fileIdToIdx = new HashMap<>();
```
done

wuwenchi commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDSimpleBucketPartitioner.java:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  Map<String, HashMap<Integer, String>> getPartitionMapper(JavaRDD<HoodieRecord<T>> records,
+                                                           HashMap<String, Integer> fileIdToIdx) {
```
done

wuwenchi commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/execution/bulkinsert/TestRDDSimpleBucketPartitioner.java:
```diff
@@ -0,0 +1,115 @@
+    JavaRDD<HoodieRecord> repartitionRecords =
+        (JavaRDD<HoodieRecord>) partitioner.repartitionRecords(HoodieJavaRDD.getJavaRDD(javaRDD), 1);
+
+    assertEquals(bucketNum * javaRDD.map(HoodieRecord::getPartitionPath).distinct().count(),
```
done

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/fs/FSUtils.java:
```diff
@@ -358,6 +358,13 @@ public static String createNewFileIdPfx() {
+  /**
+   * Returns prefix for a file group from fileId.
+   */
+  public static String getFileIdPfxFromFileId(String fileId) {
```
This looks dangerous. Instead, can you split by delimiters and take the right part.

wuwenchi replied:
done

bvaradar commented:
@wuwenchi : Please ping me in this PR once you have addressed all comments and is ready for review.

wuwenchi commented:
> @wuwenchi : Please ping me in this PR once you have addressed all comments and is ready for review.

@bvaradar All comments have now been corrected, it's ready to review now, thanks!
